### ISSUE TITLE
Notification is still active when app is not

### ISSUE THREAD

**Issue by puppykickr** [kaputnikGo/PilferShushJammer#8]
I have found a bug, I believe.
ZTE 558VL Android 7.1.1
I turned the app on, and selected 'passive'.

The notification went up, and the dialog on the screen indicated that the passive jammer was activated.

The notifications, both on the status bar and the banner, were both active.

I later went into the app, and although the notification was still up, the app dialog stated that the passive jammer was not running.

I discovered that if the App Manager is cleared, then PilferShush stops working, but the notification remains unless the app is actually force stopped.

This means that the user could believe that the device is protected even when it is not.

If the jammer(s) get turned off, I would think that the notification(s) should also go off, allowing the user to know that the app is not running.

As it is now, the user has no clue unless the app dialog is checked regularly.

I have turned battery optimization off for the app, and so it should only occur when I manually touch 'clear' on the App Manager.

Just to be clear, the App Manager that I am referring to is the one built into the device (not some app I downloaded), accessible via the right hand button on the navigation bar.

After a bit more testing, I find that if the jammer(s) are turned off manually, then the App Manager is cleared, the notifications will turn off.

But the fact remains that if the app is turned off by another means, the noifications will still be active although the jammers will not be.

**Comment by puppykickr** [comment]
My phone is the ZTE 558VL, android version 7.1.1

I use VOIP extensively, and have never had an issue with the mic not
working properly.

The issue seems to be that the notification is stating that the passive
jammer is running, but when I enter the app the dialog states that it is
not running.

This is regardless of the position of the jammer button.

In order for the app dialog to state that the jammer is working, the app
must be force stopped (at which point the notification finally disappears),
reopened, and the jammer button activated.

My biggest problem with this whole thing is that the notification clearly
states that the jammer is running, even when I confirm within the app that
it is not.

At that point the jammer on/off button has no effect on the status dialog
or the notification.

If the notification were to be useful, then it would have to vanish when
the app deactivated.

As it is now, the notification gives a false sense of security.  The user
has no idea if the jammer is really still running or not, regardless of the
presence of the notification or what it states.

I am not complaining here, I am only telling you what is happening on my
device.

Would it help to clairify the issue if I were to send a video? I should be
able to make a short video to show exactly what I am trying to describe.

Grant Ryan Swan
(puppykicker)

**Comment by puppykickr** [comment]
Also, about the issue with other apps that need the microphone from time to
time, wouldn't giving PilferShush 'Phone' permissions allow for this?  At
least the VOIP apps should work at that point, right?

Grant Ryan Swan
(puppykicker)

**Comment by kaputnikGo** [comment]
hang on, if the passive jammer is running you are still able to make VoIP calls? if this is so can you tell me which VoIP app you use so i can test against it, thanks.

**Comment by puppykickr** [comment]
There are two VOIP apps that I have used without needing to turn PilferShush off.

Telegram

Dingtone

Maybe they override any other app that has access to the microphone when they are active?

The idea that they seem to be not affected makes me wonder if PilferShush was functioning on my device after all, considering that I only recently noticed the discrepancy between the app dialog and the notification statement.

My girlfriend has an identical device, and has these same apps as well.

The other week, she called the home phone from her cell, and she could hear me but I could not hear her.

She was not using VOIP, however. She was on cellular.

**Comment by kaputnikGo** [comment]
Its quite possible that PilferShush hasnt been working properly as per your previous posts - this is hopefully fixed now with the latest update.
 
However, whether or not a VoIP app gets blocked can be down to several things: whether the VoIP app is making a phone call or a data call. 

The expected behaviour is the phone call (android.permission.CALL_PHONE) should automatically bump  PilferShush off from blocking the microphone as it is basically invoking the android phone app with " Intent callIntent = new Intent(Intent.ACTION_CALL);".  I assume this type of call would only occur if you are calling a landline or mobile phone number of someone that doesnt have the VoIP app.

The data call ( VoIP to VoIP ) should get blocked by PilferShush as it is not a phone call as such but is simply requesting microphone access as per any other user app that records audio.

How any VoIP app behaves if it doesnt get microphone access is unknown by me. But I do have the telegram app so i can test how it works.

**Comment by puppykickr** [comment]
I have not yet tested the VOIP apps with the new version, but will soon and leave a note.

The new version has cleared up the other issues that I had.

If I need to turn the app off manually to make a VOIP call, that is just fine with me, as it shows that PilferShush is indeed working.

**Comment by kaputnikGo** [comment]
Excellent, thank you again for the help with this!

**Comment by puppykickr** [comment]
OK.

Another few days, and I have discovered that after the passive jammer has
been on, the app dialog is saying that an interruption, requested by Audio
Focus, has occurred.

I thought little of this, but did notice that each time I enter the app, it
is noted again, so that multiple instances are showing.

Now today, I see this, and out of curiosity engage the active jammer.

The dialog and sounds show the active jammer to be working.

Then I disengage the passive, and the active.

The dialog and notifications respond correctly, however the sounds of the
active jammer are continuing on!

I am listening to them in amusement as I type this email.

The notification is off and the app dialog states that both jammers are off.

Just letting you know what is going on.

Grant Ryan Swan
(puppykicker)

**Comment by puppykickr** [comment]
Any time. I enjoy testing apps.

I might not be the brightest bulb in the box, but I can give feedback on real life issues from a consumer point of view.

I also am interested in simple to use security apps, such as this one.

Is there any point in time where a camera jammer is possible or could be implemented?

**Comment by kaputnikGo** [comment]
Well its very appreciated thank you. The only way for a dev to understand how their app is behaving in use is through direct feedback and the occasional crash reports that google play issue. There are so many versions of android and so many variants that there really isnt a guarantee of any function working as intended.

As for a camera jammer, i dont really know much about the camera api. The only tracker references to cameras that i have seen while researching hidden microphone use is that they can use the front facing camera to take photos of the user. The other use i have seen is with scanning the photos folder for filenames, dates and numbers etc. 

There maybe something on f-droid.org that is suitable. otherwise put a sticker over the front facing camera(s) :)

**Comment by kaputnikGo** [comment]
Actually, this: https://developer.android.com/guide/topics/admin/device-admin
leads to this: https://developer.android.com/reference/android/app/admin/DevicePolicyManager

which has a
 public void setCameraDisabled (ComponentName admin, boolean disabled)

**Comment by puppykickr** [comment]
Thank you very much.

Although there really wasnt anything there I could use, I saw 'device
administrator' and went on a search.

From there I found several apps that have device administer rights over the
camera.

Unfortunately, none on F-Droid.

I found two that I like well, and am trying them out.

I have a nifty little app from F-Droid called Microphone, which actually
uses the microphone on your phone like one on a PA system.

It give feedback readilly on the built in speaker, but works just fine
through bluetooth, or any other system that allows distance between the
microphone and the speaker.

So far, it appears that PilferShush is working.

If PilferShush is engaged (passive) the microphone is not working.

If Microphone app is engaged first, then PilferShush, I get a dialog
stating something to the effect of 'Passive jammer not initializing,
possibly because the microphone is in use.'

I think this is how you want it to perform, if I am correct.

I checked these camera apps the same way, both with the built in camera app
and other apps I have that use the camera.

Maybe it is possible to integrate a device administrator for both the
camera and the microphone.

A notification, and possibly an icon or two in the quick settings, would be
a very convenient way for the user to access control when necessary.

Administrator control would also stop your app from being shut down by the
device, allowing it to run continuously, if I am not mistaken, correct?

It seems that this could solve some issues.

There are a few apps on the PlayStore that have both camera and microphone
blocking, but they have ads or require payment.

Your app seems to be the only microphone jammer on F-Droid, and I found no
camera blockers at all.

If your app were to have both of these options yours would be the only one
of its kind on F-Droid.

Grant Ryan Swan
(puppykicker)
